### DISCUSSION TITLE
Trying to containerize Trame application - getting [Errno 99] error while attempting to bind on address ('::1', 8080, 0, 0)

### DISCUSSION THREAD

**Discussion by charlycou** [Kitware/trame#564] [Q&A]
Hi,
I'm trying to containerize a trame app but I'm getting  `[Errno 99] error while attempting to bind on address ('::1', 8080, 0, 0): cannot assign requested address` when trying to access to the served content.
I checked that the port is actually free using `lsof -i :8080` and I tried to start the server using two other random ports (8181 and 27000) but I'm getting the same exception.

I created a repo to reproduce: https://github.com/charlycou/test-trame-geovizpy

```
docker build -t trame-test-img .
docker run -it -d --name trame-test -p 80:80 trame-test-img
```
And open the application at localhost:80. Then check the logs.

```
docker exec -it trame-test /bin/bash
```
Logs are in `/deploy/server/logs/launcher` directory.

**Does anyone have any ideas for solving this problem?** I successfully run the example from the doc that use the same base image so it must be due to geovizpy.py

Cheers,

**Comment by jourdain** [comment]
https://github.com/Kitware/trame-server/blob/master/trame_server/core.py#L556

**Reply by charlycou** [reply]
Thank you for pointing me out the configuration parameter that will solve my issue. Could you elaborate further? I don't understand what's wrong starting the server on "localhost" with port 8080 if I'm forwarding this port running the container. 

Also, I realize that `server.start()` function used in[ the basic example](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/SingleFile) start the server  on "localhost" with port 8080 (default config) while the port forwarded to serve the client is 80 (`docker run -it --rm -p 8080:80 trame-app`). Could you tell me what I'm missing?

Thank you for your feedback, I'm certainly a bit confused.

**Reply by jourdain** [reply]
Set it to 127.0.0.1. That will fix the IP resolution as localhost is missing a mapping within the docker image.

**Comment by EZoni** [comment]
Hi,

I'm refreshing this discussion because I might be encountering a similar issue.

I have a Trame application that runs successfully on my computer and opens a browser window with a GUI when I run it manually.

The application is very simple, it displays a few text fields, buttons, and plotly-based line plots.

I am now setting up a Dockerfile to build an image and run the application from a Docker container, which is basically as follows:
```docker
# Use an official Python runtime as a parent image
FROM continuumio/miniconda3:latest

# Set the working directory in the container
WORKDIR /app

# Copy the current directory contents into the container at /app
COPY . /app

# Install any needed packages specified in environment.yml
RUN conda env create -f gui.yml

# Make port 8080 available to the world outside this container
EXPOSE 8080

# Set environment variable to run GUI as desktop application
ENV PYWEBVIEW_GUI=qt

# NOTE Need to activate conda environment at container launch
# Run app.py when the container launches
CMD ["conda", "run", "-n", "gui", "python", "app.py", "--port", "8080"]
```

When I build the image and run the container 
```bash
sudo docker build -t gui-docker-image .
sudo docker run -p 8080:8080 gui-docker-image
```
the application runs without crashing, but it does not open up a browser window with the GUI.

This is the output when I run the container:
```console
...

And to prevent your browser from opening, add '--server' to your command line.

App running at:
 - Local:   http://localhost:8080/
 - Network: http://127.0.0.1:8080/

Note that for multi-users you need to use and configure a launcher.
```

Usually at this point the terminal is hanging, the browser window is open, and I have to close the port with `Ctrl`+`C`.

This time the terminal is not hanging and the browser window does not open.

What am I missing?

From the answer above,
> Set it to 127.0.0.1. That will fix the IP resolution as localhost is missing a mapping within the docker image.

I did not understand what I would need to change to make it work, what should be set to 127.0.0.1 in my Dockerfile and/or commands I use to build the image and run the container.

Thanks in advance for your help.

**Reply by jourdain** [reply]
Can you explain your intent on why you want to use docker?  (dev, multi-user, else?)

Also are you expecting the browser to open within docker or on your host system?

**Reply by EZoni** [reply]
Sure, thanks.

We are trying to put together some sort of super-facility infrastructure that orchestrates simulations, ML models, and experiments.

> Can you explain your intent on why you want to use docker? (dev, multi-user, else?)

At this point we think multi-user is not necessarily important to us.

> Also are you expecting the browser to open within docker or on your host system?

As for the browser, I think we would like to achieve the following:
- the browser runs on a computer outside the Docker container;
- the Trame app runs within the Docker container on a NERSC node (through [Spin](https://www.nersc.gov/systems/spin/)).

The fact that the browser "opens up" automatically is not really important to us in this context. It would be enough that we can click on the URL displayed in the terminal output (there are two URLs, labeled "Local" and "Network") when the Trame app runs and that the URL brings us to the GUI. The URL does not seem to work at the moment.

**Reply by jourdain** [reply]
Thanks for those add-on information, but you are still not explaining why you are using docker. Is it to encapsulate the application for 1 user?
Also you need to point your browser to the IP on where the container is running. If you use `localhost` or `127.0.0.1`, those will just connect to your local machine, not the node from your cluster.

**Reply by jourdain** [reply]
In term of add-on args you should add `--host 0.0.0.0 --server`

**Reply by EZoni** [reply]
Thanks, here are some more details.

> Thanks for those add-on information, but you are still not explaining why you are using docker. Is it to encapsulate the application for 1 user?

Yes, I think we want to encapsulate the application for our users (experimentalists). Containerization also seems the natural thing to do when relying on the Spin infrastructure (slide from NERSC's Spin tutorial):

![Screenshot from 2025-01-13 10-17-18](https://github.com/user-attachments/assets/b43e339a-996d-4047-9948-7be305281dcf)

> Also you need to point your browser to the IP on where the container is running. If you use localhost or 127.0.0.1, those will just connect to your local machine, not the node from your cluster.

Right now I'm at the "build" stage with respect to the figure above. I'm running everything locally, on my computer, no cluster involved, and I'm trying to test my containerization. I'm trying to understand why, locally, those URLs do not work, even though the app runs without crashing, and what I could do to fix it, given that when I run the app manually, without the container, everything works.

**Reply by jourdain** [reply]
It should work unless you use remote rendering but that should also trigger an error on rendering.

First, make sure you fix the network part, then tackle the rendering part.

**Reply by EZoni** [reply]
Thanks, I think it works now.

Just for clarity, I modified the Dockerfile by adding `--host 0.0.0.0 --server` to the Trame app launch command:
```docker
# Use an official Python runtime as a parent image
FROM continuumio/miniconda3:latest

# Set the working directory in the container
WORKDIR /app

# Copy the current directory contents into the container at /app
COPY . /app

# Install any needed packages specified in environment.yml
RUN conda env create -f gui.yml

# Make port 8080 available to the world outside this container
EXPOSE 8080

# Set environment variable to run GUI as desktop application
ENV PYWEBVIEW_GUI=qt

# NOTE Need to activate conda environment at container launch
# Run app.py when the container launches
CMD ["conda", "run", "-n", "gui", "python", "app.py", "--port", "8080", "--host", "0.0.0.0", "--server"]
```

With this Dockerfile, I build the image and run the container as described in my first message (no changes):
```bash
sudo docker build -t gui-docker-image .
sudo docker run -p 8080:8080 gui-docker-image
```

Once this is done, I can navigate to http://localhost:8080/ and see the GUI.

My mistakes were:
1. I was missing `--host 0.0.0.0 --server` in the Trame app launch command set in the Dockerfile;
2. After fixing 1., I was navigating locally to http://0.0.0.0:8080/ instead of http://localhost:8080/. I was misled by the terminal output of the Trame app (probably due to lack of understanding on my end, although I would expect Trame to print out the correct URLs):
```console
App running at:
 - Local:   http://0.0.0.0:8080/
 - Network: http://0.0.0.0:8080/
```

**Reply by jourdain** [reply]
The reason why trame can't print localhost is because the docker image is missing a name for its host, and only the provided `0.0.0.0` is available to the network layer. Just some network stuff on a bare bone linux setup issue.
